### PR TITLE
feat: [IDP-2302] management payment authorization with etag

### PR DESCRIPTION
**Pull request by Giuseppe-LaManna**
#### Description

[IDP-2302](https://pagopa.atlassian.net/browse/IDP-2302)
#### List of Changes

- Added If-Match Header in authorization invocation
- Added new check in the common auth service
- Updated tests
#### How Has This Been Tested?

- Pre-Deploy Test
    - [x] Unit
    - [ ] Integration (Narrow)
- Post-Deploy Test
    - [ ] Isolated Microservice
    - [ ] Broader Integration
    - [ ] Acceptance
    - [ ] Performance & Load

#### Types of changes

- [ ] PATCH - Bug fix (backwards compatible bug fixes)
- [x] MINOR - New feature (add functionality in a backwards compatible manner)
- [ ] MAJOR - Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

[IDP-2302]: https://pagopa.atlassian.net/browse/IDP-2302?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/it/gov/pagopa/payment/connector/rest/reward/mapper/RewardCalculatorMapper.java [hunk 1]
```diff
@@ -32,6 +33,11 @@ public AuthPaymentRequestDTO rewardMap(TransactionInProgress transactionInProgre
                 .channel(transactionInProgress.getChannel())
                 .build();
     }
+    public AuthPaymentRequestDTO authRequestMap(TransactionInProgress transactionInProgress) {
+        return AuthPaymentRequestDTO.builder()
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/connector/rest/reward/mapper/RewardCalculatorMapper.java:37]: this body is totally empty except the rewardCents field!
you could write a common method to call in order to fill the model in both methods...

File: src/main/java/it/gov/pagopa/payment/repository/TransactionInProgressRepositoryExtImpl.java [hunk 2]
```diff
@@ -158,11 +159,11 @@ public void updateTrxRelateUserIdentified(String id, String userId, String chann
                 TransactionInProgress.class);
     }
     @Override
-    public void updateTrxIdentified(String id, String userId, Long reward, List<String> rejectionReasons, Map<String, List<String>> initiativeRejectionReasons, Map<String, Reward> rewards, String channel) {
+    public void updateTrxWithStatus(String id, String userId, Long reward, List<String> rejectionReasons, Map<String, List<String>> initiativeRejectionReasons, Map<String, Reward> rewards, String channel, SyncTrxStatus status) {
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/repository/TransactionInProgressRepositoryExtImpl.java:162]: you should also accept and save the retrieved  counterVersion

File: src/main/java/it/gov/pagopa/payment/service/payment/common/CommonAuthServiceImpl.java [hunk 3]
```diff
@@ -117,6 +122,15 @@ protected void checkAuth(String trxCode, TransactionInProgress trx){
 
     }
 
+    protected void checkTrxStatusToInvokePreAuth(String userId, SyncTrxStatus status, TransactionInProgress trx) {
+        if ((status.equals(SyncTrxStatus.CREATED) && userId != null) ||
+                (status.equals(SyncTrxStatus.IDENTIFIED) && trx.getReward() == null)){
+           commonPreAuthService.previewPayment(trx,trx.getChannel(),SyncTrxStatus.AUTHORIZATION_REQUESTED);
+        }else if(!LIST_OF_STATUS_TO_IGNORE.contains(status)) {
+            trx.setStatus(SyncTrxStatus.AUTHORIZATION_REQUESTED);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/service/payment/common/CommonAuthServiceImpl.java:130]: you have to store the status inside the database before to invoke rewardCalculator

File: src/main/java/it/gov/pagopa/payment/service/payment/common/CommonAuthServiceImpl.java
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/payment/service/payment/common/CommonAuthServiceImpl.java:1]: this setting should not more executed inside the auth behavior:
trx.setTrxChargeDate(OffsetDateTime.now().truncatedTo(ChronoUnit.MILLIS));

Comment (antonioT90) [reply src/main/java/it/gov/pagopa/payment/service/payment/common/CommonAuthServiceImpl.java:1]: MoreOver, the authorization now should save just the status! The reward now is stored by the preview

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/payment/connector/rest/reward/RewardCalculatorRestClient.java [hunk 4]
```diff
@@ -17,14 +19,15 @@ public interface RewardCalculatorRestClient {
           produces = MediaType.APPLICATION_JSON_VALUE)
   @ResponseBody
   ResponseEntity<AuthPaymentResponseDTO> previewTransaction(@PathVariable("initiativeId") String initiativeId,
-                                                           @RequestBody AuthPaymentRequestDTO body);
+                                                           @RequestBody PreAuthPaymentRequestDTO body);
```
Comment (petretiandrea) [inline src/main/java/it/gov/pagopa/payment/connector/rest/reward/RewardCalculatorRestClient.java:22]: I think we could use a more expressive name. We don't have a real pre-authorization phase, so we could refer to this DTO as a payment request, so `PaymentRequestDTO` I think is more appropriated

**Pull request merged**